### DISCUSSION TITLE
Global exception handling in cli

### DISCUSSION THREAD

**Discussion by nils-a** [spectreconsole/spectre.console#601] [Q&A] [answered]
I have had some problems in one of my programs which made me regret that I didn't implement a global catch handler to log my exceptions.
In the end, I wrote something along the lines of:

```cs
using System.Diagnostics.CodeAnalysis;
using Spectre.Console;
using Spectre.Console.Cli;

var app = new CommandApp<BrokenCommand>();
app.Configure(c =>
{
    c.PropagateExceptions();
});

try
{
    return app.Run(args);
}
catch (Exception e)
{
    // imagine fancy logging here.
    
    // now create some output, because PropagateExceptions() "swallows" the normal output.
    AnsiConsole.WriteException(e, ExceptionFormats.ShortenEverything);
    return -99;
}

internal sealed class BrokenCommand : Command<BrokenCommand.Settings>
{
    internal class Settings : CommandSettings
    {
    }

    public override int Execute([NotNull] CommandContext context, [NotNull] Settings settings)
    {
        throw new ApplicationException("Boom!");
        return 0;
    }
}
```

Now, this code has two major drawbacks:
* I need to set `PropagateExceptions` in addition to writing the `try`-`catch` in order to be able to handle exceptions.
* Setting `PropagateExceptions` removes the default output of "Error: [...something happened...]" so I needed to add some additional output.

So, the first question is "Am I missing something"? 
If not, would someone be interested in having a bit more error handling to the cli?

Something like `AddExceptionHandler` that will get an `Action<Exception>` to do custom stuff with it - like logging.
Additionally, I could imagine having a `ExceptionRenderer` that outputs a different rendering of a final exception and a `ExceptionReturnCode` that will let you customize the return code when an exception occurs. So my example from above could potentially look like:

```cs
var app = new CommandApp<BrokenCommand>();
app.Configure(c =>
{
    c.AddExceptionHandler(ex => 
    {
      // add your cool logging here
    });
    c.SetExceptionRenderer(ex => 
    {
        AnsiConsole.WriteException(e, ExceptionFormats.ShortenEverything);
    });with
    c.SetExceptionReturnCode(-99);
});

return app.Run(args);
```

Which I feel would be much cleaner than my code above. Would changes like those make sense for someone?

**Comment by patriksvensson** [comment]
I've been thinking about this as well.

It might be a good idea to have a single method that handles the exception and also returns an exit code.

**Reply by nils-a** [reply]
Something like:

```cs
var app = new CommandApp<BrokenCommand>();
app.Configure(c =>
{
    c.HandleException(ex => 
    {
        // add your cool logging here
        AnsiConsole.WriteException(e, ExceptionFormats.ShortenEverything);
        return -99;
    });
});

return app.Run(args);
```

**Reply by patriksvensson** [reply]
Yes, but perhaps it's nice to have some kind of default exception rendering. The question is if it's possible to solve without having a really complicated API. I assume that writing something to the console is part of handling the exception.

I don't think it's possible to have two overloads diffing by return type (`Action<Exception>` and `Func<Exception, int>`), so the best idea is probably to only have one method where you do everything yourself.

**Reply by nils-a** [reply]
> I assume that writing something to the console is part of handling the exception.

Actually I only wanted to log, the rest was fine for me. 😉 

I think we could use two extensions, one taking a `Action<Exception>` and one taking `Func<Exception, int>`, additionally we could add a bool to suppress the default console output, if one would use a custom output.

So, something like 
```cs
var app = new CommandApp<BrokenCommand>();
app.Configure(c =>
{
    // either using Action<Exception>
    c.HandleException(ex => 
    {
        // logging here
    });

    // or Func<Exception, int>
    c.HandleException(ex => 
    {
        // logging here
        return -99;
    });

    // or adding custom output and set suppressDefaultOutput to true
    c.HandleException(ex => 
    {
        AnsiConsole.WriteException(e, ExceptionFormats.ShortenEverything);
    }, suppressDefaultOutput: true);

    // last option: supply the exit code and set suppressDefaultOutput to true
    c.HandleException(ex => 
    {
        AnsiConsole.WriteException(e, ExceptionFormats.ShortenEverything);
        return -99;
    }, suppressDefaultOutput: true);
});

return app.Run(args);
``` 

I *think* an API like that could be possible.

**Reply by patriksvensson** [reply]
Hmm, I like the first two extensions but I'm a bit hesitant about the boolean parameter, to be honest.

**Reply by nils-a** [reply]
> I'm a bit hesitant about the boolean parameter

I thought so :smile:
I think I personally would have chosen the three methods for the three "functionalities" (setting exit code / modifying the output / do something with the exception) but I feel having a consistent API should be our primary goal. As long as it's documented, it's fine for me.

I'll probably try to code something in the evening, today.

**Comment by nils-a** [comment, marked as answer]
Converted to https://github.com/spectreconsole/spectre.console/issues/606
